### PR TITLE
Combining Index #399

### PR DESCRIPTION
**Pull request by LuciferYang**
## What changes were proposed in this pull request?

1、IndexContext: getBestIndexer -> getAvailableIndexers to choose high priority indexers and a parameter named maxChooseSize to  available indexers max length.
2、add IndexerScanners to wrap Seq[Indexer] and act their operations
3、IndexContext hold IndexerScanners instead of IndexerScanner
4、IndexContext: buildScanner-> buildScanners to build a IndexerScanners instance
5、OapDataReader： filterScanner->filterScanners
6、SQLConf： add new conf "spark.sql.oap.indexer.max.use.size"
7、add CombiningIndexSuite to test Combining Index
8、add a new test case in IndexSelectionSuite to test getAvailableIndexers
9、fix other test case semantics 

## How was this patch tested?
1、existing case
2、IndexSelectionSuite and CombiningIndexSuite

### CODE REVIEW

**Comment by LuciferYang** [comment]
@adrian-wang @linhong-intel @xuanyuanking @yueawang

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala [hunk 1]
```diff
@@ -243,10 +244,44 @@ private[oap] object ScannerBuilder extends Logging {
...
+  def order: SortDirection = actualUsedScanners.head.meta.indexType.indexOrder.head
+
+  def initialize(dataPath: Path, conf: Configuration): IndexScanners = {
+    actualUsedScanners.map(_.initialize(dataPath, conf))
+    backend = actualUsedScanners.map(_.toArray)
+      .sortBy(_.length)
+      .reduce((left, right) => {
+        if (left.isEmpty) left
+        else left.intersect(right)
+      }).toIterator
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: should parallelizable to Array when actualUsedScanners.length > 1 like this:
 backend = actualUsedScanners.length match {
      case 0 => Iterator.empty
      case 1 => actualUsedScanners.head.toArray.toIterator
      case _ => actualUsedScanners.par.map(_.toArray).seq
        .sortBy(_.length)
        .reduce((left, right) => {
          if (left.isEmpty) left
          else left.intersect(right)
        }).toIterator
    }

**Comment by adrian-wang** [comment]
Thanks for working on this, I'll merge this by today.

File: src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -728,6 +728,13 @@ object SQLConf {
       .doubleConf
       .createWithDefault(0.7)
 
+  val OAP_INDEXER_CHOICE_MAX_SIZE =
+    SQLConfigBuilder("spark.sql.oap.indexer.max.use.size")
+      .internal()
+      .doc("The max availabe indexer choose size.")
+      .intConf
+      .createWithDefault(2)
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: how about changing it to 1 so this becomes an advanced feature? we can set it to 2 when the benchmark shows good results.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala [hunk 3]
```diff
@@ -109,53 +109,108 @@ private[oap] class IndexContext(meta: DataSourceMeta) extends Logging {
...
-    var lastIdx = -1
-    var bestIndexer: IndexMeta = null
-    var ratio: Double = 0.0
-    var isFirst = true
-    logDebug("Get Best Index:")
-    for ((idx, indexMeta) <- availableIndexes) {
+  def getAvailableIndexers(attrNum: Int, maxChooseSize: Int = 2): Seq[(Int, IndexMeta)] = {
+    logDebug("Get Available Indexers: maxChooseSize = " + maxChooseSize)
+
+    def takeRatioAndUsedFields(attrNum: Int,
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: Nit: format should looks like
```
def funcNameVeryLong(
[4spaces]param1: Type,
[4spaces]param2: Type): ReturnType = ???
```

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala [hunk 4]
```diff
@@ -277,7 +277,8 @@ private[oap] object ScannerBuilder extends Logging {
   }
 
   def build(filters: Array[Filter], ic: IndexContext,
-      scannerOptions: Map[String, String] = Map.empty): Array[Filter] = {
+            maxChooseSize: Int = 2,
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: also style

File: src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: ok~

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala (hunk 3 above)
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: ok~

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala (hunk 4 above)
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: ok~

**Comment by LuciferYang** [comment]
@adrian-wang thx~

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala [hunk 5]
```diff
@@ -293,10 +296,51 @@ private[oap] object ScannerBuilder extends Logging {
...
+  def initialize(dataPath: Path, conf: Configuration): IndexScanners = {
+    backendIter = actualUsedScanners.length match {
+      case 0 => Iterator.empty
+      case 1 =>
+        actualUsedScanners.head.initialize(dataPath, conf)
+        actualUsedScanners.head.toArray.iterator
+      case _ =>
+        actualUsedScanners.par.foreach(_.initialize(dataPath, conf))
+        actualUsedScanners.map(_.toArray)
+          .sortBy(_.length)
```
Comment (yueawang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: Is this sort necessary?? 
There is a potential issue that sort changes the order of the returned rowids.
For example
a = (1, 2, 3, 4, 5)
b = (5, 4, 3)
a.intersect(b) => (3, 4, 5) // without sort, asc order.
b.intersect(a) => (5, 4, 3) // with sort, desc order.
Now it's OK as optimized strategies assume only 1 index, but it potentially causes a problem if we take advantage of the order of the index.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala [hunk 6]
```diff
@@ -293,10 +296,51 @@ private[oap] object ScannerBuilder extends Logging {
...
+          .reduce((left, right) => {
+            if (left.isEmpty) left
+            else left.intersect(right)
```
Comment (yueawang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: The intersect is a AND operator, will it be extended to OR, and maybe more other logical operators in future?

Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: I think If 'a=va or b=vb' can use index and  we can keep the relationship of Seq[IndexScanner] like '(scannerA OR scannerB) ', there should make corresponding optimization.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala (hunk 5 above)
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: maybe unecessary...., should delete it~ but for multi scanners which scanner's order should be finally used?

Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: do sortBy(_.length) is use for "if (left.isEmpty) left".
For example
a = (1, 2, 3, 4, 5)
b = (5, 4, 3)
c =()
without sort, we shoud do a.intersect(b).intersect(c)，but with with sort can bring a little optimization, but i'm not sure should use left.intersect(right) or right.intersect(left) .....
Now, for multi scanners which scanner's order should be finally used?

Comment (yueawang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: For multi scanner, so far we don't ask for any order, any order is acceptable. However if later we take advantage of an specific order, e.g. a query with condition like "where a > 5 and b < 100 order by a", taking the order from index of column a should be better as we may skip the potential sort. In that case, I think this sort can be removed.

Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala]: This case is complex，we can take advantage of an specific order depending on at least two conditions:
1、Available Indexers Seq[IndexScanner] is (IndexScannerA, IndexScannerB) not (IndexScannerB, IndexScannerA)
2、IndexScannerA not filter out by indexIsAvailable method

But for this potential advantage i will get a pr to remove it ~This case is complex，we can take advantage of an specific order depending on at least two conditions:
1、Available Indexers Seq[IndexScanner] is (IndexScannerA, IndexScannerB) not (IndexScannerB, IndexScannerA)
2、IndexScannerA not filter out by indexIsAvailable method

But for this potential advantage i will get a pr to remove it ~

**Pull request merged**